### PR TITLE
Split SpirvLowerAlgebraTransform pass into two passes. NFC.

### PR DESCRIPTION
**Pull request by perlfu**
SpirvLowerAlgebraTransform is invoked twice with two distinct
functions. Refactor it into two separate passes which share
a common base class.

### CODE REVIEW

**Comment by amdvlk-admin** [comment]
### Test summary for commit ce3f3cb3ae3fa8e8a4b2fcbc52f1d4594357b23d
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 712f76fee75d69b23a1ea8f6465752c3ccaaf9a2</li>
<li>PAL: dev d128e79bf98557761304e6aa46ba143fffb551d8</li>
<li>SPVGEN: dev 6205d0b42ebec5c8c01ff41e97fe51b56aa67343</li>
<li>XGL: dev d34fcbf7a0a5755cbac7b27c5784b7ed6782e4c2</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev b3a0782762d195b57a2bebaabca24be90a35db2a</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix)</summary><ul>


<details open><summary>Rhel 8.2, Gfx8</summary><ul>

<li>Passed: 142044/142044 (100.0%)</li>
<li>Failed: 0/142044 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>No Results</li>

</ul></details>


</ul></details>

**Comment by JaxLinAMD** [comment]
retest this please

**Comment by trenouf** [comment]
The new files have the wrong copyright header.

An alternative approach would be to keep everything in the same source file as it was before, but have two different passes, sharing some of the same code (or a base class as you have here). These are passes, not analyses, so the class declaration of the pass does not need to be in a .h file. The technique used in LLVM is to put the class definition in the .cpp file in the anonymous namespace. See VertexFetch.cpp for an example of where I have done that in LGC.

**Comment by perlfu** [comment]
Would you prefer these in a single source file then?

Also by wrong copyright, do you mean wrong year since I copied original source year?

**Comment by JaxLinAMD** [comment]
retest this please

**Comment by JaxLinAMD** [comment]
retest this please

**Comment by JaxLinAMD** [comment]
retest this please

**Comment by JaxLinAMD** [comment]
retest this please

**Comment by amdvlk-admin** [comment]
### Test summary for commit ce3f3cb3ae3fa8e8a4b2fcbc52f1d4594357b23d
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 712f76fee75d69b23a1ea8f6465752c3ccaaf9a2</li>
<li>PAL: dev d128e79bf98557761304e6aa46ba143fffb551d8</li>
<li>SPVGEN: dev 6205d0b42ebec5c8c01ff41e97fe51b56aa67343</li>
<li>XGL: dev d34fcbf7a0a5755cbac7b27c5784b7ed6782e4c2</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev b3a0782762d195b57a2bebaabca24be90a35db2a</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix)</summary><ul>


<details open><summary>Rhel 8.2, Gfx8</summary><ul>

<li>Passed: 142044/142044 (100.0%)</li>
<li>Failed: 0/142044 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 142088/142088 (100.0%)</li>
<li>Failed: 0/142088 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by JaxLinAMD** [comment]
retest this please

**Comment by amdvlk-admin** [comment]
### Test summary for commit ce3f3cb3ae3fa8e8a4b2fcbc52f1d4594357b23d
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 712f76fee75d69b23a1ea8f6465752c3ccaaf9a2</li>
<li>PAL: dev d128e79bf98557761304e6aa46ba143fffb551d8</li>
<li>SPVGEN: dev 6205d0b42ebec5c8c01ff41e97fe51b56aa67343</li>
<li>XGL: dev d34fcbf7a0a5755cbac7b27c5784b7ed6782e4c2</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev b3a0782762d195b57a2bebaabca24be90a35db2a</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix)</summary><ul>


<details open><summary>Rhel 8.2, Gfx8</summary><ul>

<li>No Results</li>

</ul></details>

<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>No Results</li>

</ul></details>


</ul></details>

**Comment by amdvlk-admin** [comment]
### Test summary for commit ce3f3cb3ae3fa8e8a4b2fcbc52f1d4594357b23d
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 712f76fee75d69b23a1ea8f6465752c3ccaaf9a2</li>
<li>PAL: dev d128e79bf98557761304e6aa46ba143fffb551d8</li>
<li>SPVGEN: dev 6205d0b42ebec5c8c01ff41e97fe51b56aa67343</li>
<li>XGL: dev d34fcbf7a0a5755cbac7b27c5784b7ed6782e4c2</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev b3a0782762d195b57a2bebaabca24be90a35db2a</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix)</summary><ul>


<details open><summary>Rhel 8.2, Gfx8</summary><ul>

<li>Passed: 142044/142044 (100.0%)</li>
<li>Failed: 0/142044 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 142088/142088 (100.0%)</li>
<li>Failed: 0/142088 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 142138/142138 (100.0%)</li>
<li>Failed: 0/142138 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by perlfu** [comment]
I have reworked this into a single file with two classes sharing a common base.
Classes have been renamed as per Rex's suggestion.

**Comment by amdvlk-admin** [comment]
### Test summary for commit 110aaa971a02053a8be94b133c3d7912de97ded2
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 712f76fee75d69b23a1ea8f6465752c3ccaaf9a2</li>
<li>PAL: dev d128e79bf98557761304e6aa46ba143fffb551d8</li>
<li>SPVGEN: dev 6205d0b42ebec5c8c01ff41e97fe51b56aa67343</li>
<li>XGL: dev d34fcbf7a0a5755cbac7b27c5784b7ed6782e4c2</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev b3a0782762d195b57a2bebaabca24be90a35db2a</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix)</summary><ul>


<details open><summary>Rhel 8.2, Gfx8</summary><ul>

<li>Passed: 142044/142044 (100.0%)</li>
<li>Failed: 0/142044 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 142088/142088 (100.0%)</li>
<li>Failed: 0/142088 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 142138/142138 (100.0%)</li>
<li>Failed: 0/142138 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by amdvlk-admin** [comment]
### Test summary for commit 6c9ed6d67a40434a09dfd3c8f6abb675fac68989
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 712f76fee75d69b23a1ea8f6465752c3ccaaf9a2</li>
<li>PAL: dev d128e79bf98557761304e6aa46ba143fffb551d8</li>
<li>SPVGEN: dev 6205d0b42ebec5c8c01ff41e97fe51b56aa67343</li>
<li>XGL: dev d34fcbf7a0a5755cbac7b27c5784b7ed6782e4c2</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev b3a0782762d195b57a2bebaabca24be90a35db2a</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/426270)</summary><ul><li>Built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix</li>
<details open><summary>Rhel 8.2, Gfx8</summary><ul>

<li>Passed: 142044/142044 (100.0%)</li>
<li>Failed: 0/142044 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 142088/142088 (100.0%)</li>
<li>Failed: 0/142088 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 142138/142138 (100.0%)</li>
<li>Failed: 0/142138 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by amdvlk-admin** [comment]
### Test summary for commit 23dce93e36803a3221a9793582e27917a4725342
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 712f76fee75d69b23a1ea8f6465752c3ccaaf9a2</li>
<li>PAL: dev d128e79bf98557761304e6aa46ba143fffb551d8</li>
<li>SPVGEN: dev 6205d0b42ebec5c8c01ff41e97fe51b56aa67343</li>
<li>XGL: dev d34fcbf7a0a5755cbac7b27c5784b7ed6782e4c2</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev b3a0782762d195b57a2bebaabca24be90a35db2a</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/426270)</summary><ul><li>Built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix</li>
<details open><summary>Rhel 8.2, Gfx8</summary><ul>

<li>Passed: 142044/142044 (100.0%)</li>
<li>Failed: 0/142044 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 142088/142088 (100.0%)</li>
<li>Failed: 0/142088 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 142138/142138 (100.0%)</li>
<li>Failed: 0/142138 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by amdvlk-admin** [comment]
### Test summary for commit a1c6818fd823348dbdf83896972fc58dfb8dbbfd
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 712f76fee75d69b23a1ea8f6465752c3ccaaf9a2</li>
<li>PAL: dev d128e79bf98557761304e6aa46ba143fffb551d8</li>
<li>SPVGEN: dev 6205d0b42ebec5c8c01ff41e97fe51b56aa67343</li>
<li>XGL: dev d34fcbf7a0a5755cbac7b27c5784b7ed6782e4c2</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev b3a0782762d195b57a2bebaabca24be90a35db2a</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/426270)</summary><ul><li>Built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix</li>
<details open><summary>Rhel 8.2, Gfx8</summary><ul>

<li>Passed: 142044/142044 (100.0%)</li>
<li>Failed: 0/142044 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 142088/142088 (100.0%)</li>
<li>Failed: 0/142088 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 142138/142138 (100.0%)</li>
<li>Failed: 0/142138 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by amdvlk-admin** [comment]
### Test summary for commit 157abdb7e59eb29f780fd42b4d8ca2c20e428a34
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 712f76fee75d69b23a1ea8f6465752c3ccaaf9a2</li>
<li>PAL: dev d128e79bf98557761304e6aa46ba143fffb551d8</li>
<li>SPVGEN: dev 6205d0b42ebec5c8c01ff41e97fe51b56aa67343</li>
<li>XGL: dev d34fcbf7a0a5755cbac7b27c5784b7ed6782e4c2</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev b3a0782762d195b57a2bebaabca24be90a35db2a</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/426270)</summary><ul><li>Built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix</li>
<details open><summary>Rhel 8.2, Gfx8</summary><ul>

<li>Passed: 142044/142044 (100.0%)</li>
<li>Failed: 0/142044 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 142088/142088 (100.0%)</li>
<li>Failed: 0/142088 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 142138/142138 (100.0%)</li>
<li>Failed: 0/142138 (0.0%)</li>
</ul></details>


</ul></details>

File: llpc/lower/llpcSpirvLowerMath.cpp [hunk 1]
```diff
@@ -24,60 +24,110 @@
...
+#include "llvm/IR/InstVisitor.h"
 #include "llvm/IR/Instructions.h"
 #include "llvm/IR/Operator.h"
 #include "llvm/Support/Debug.h"
 #include "llvm/Support/raw_ostream.h"
 #include "llvm/Transforms/Utils.h"
 #include "llvm/Transforms/Utils/Local.h"
 
-#define DEBUG_TYPE "llpc-spirv-lower-algebra-transform"
+#define CONST_FOLDING_DEBUG_TYPE "llpc-spirv-lower-math-const-folding"
```
Comment (amdrexu) [inline llpc/lower/llpcSpirvLowerMath.cpp:48]: CONST_FOLDING_DEBUG_TYPE -> DEBUG_TYPE_CONST_FOLDING

File: llpc/lower/llpcSpirvLowerMath.cpp [hunk 2]
```diff
@@ -24,60 +24,110 @@
...
+#define FLOAT_OP_DEBUG_TYPE "llpc-spirv-lower-math-float-op"
```
Comment (amdrexu) [inline llpc/lower/llpcSpirvLowerMath.cpp:49]: FLOAT_OP_DEBUG_TYPE -> DEBUG_TYPE_FLOAT_OP

File: llpc/lower/llpcSpirvLowerMath.cpp [hunk 3]
```diff
@@ -24,60 +24,110 @@
...
+
+  void flushDenormIfNeeded(llvm::Instruction *inst);
+  bool isOperandNoContract(llvm::Value *operand);
+  void disableFastMath(llvm::Value *value);
+
+  bool m_changed;         // Whether the module is changed
+  bool m_fp16DenormFlush; // Whether FP mode wants f16 denorms to be flushed to zero
+  bool m_fp32DenormFlush; // Whether FP mode wants f32 denorms to be flushed to zero
+  bool m_fp64DenormFlush; // Whether FP mode wants f64 denorms to be flushed to zero
+  bool m_fp16Rtz;         // Whether FP mode wants f16 round-to-zero
```
Comment (amdrexu) [inline llpc/lower/llpcSpirvLowerMath.cpp:76]: m_fp16Rtz -> m_fp16RoundToZero

**Comment by amdvlk-admin** [comment]
### Test summary for commit ce3f3cb3ae3fa8e8a4b2fcbc52f1d4594357b23d
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 712f76fee75d69b23a1ea8f6465752c3ccaaf9a2</li>
<li>PAL: dev b30561984a7b697ed7c709be76ff2e4d30532728</li>
<li>SPVGEN: dev b28e58f562591f335b9a3b5ae7e578a426df210b</li>
<li>XGL: dev 9bcc23fb29dab0b7ed1ebff17f3e99ae4da0bdf9</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev b3a0782762d195b57a2bebaabca24be90a35db2a</li>
</ul></details>

<details><summary>CTS tests (Failed: 1/441732)</summary><ul><li>Built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix</li>
<details open><summary>Rhel 8.2, Gfx8</summary><ul>

<li>Passed: 147198/147198 (100.0%)</li>
<li>Failed: 0/147198 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 147242/147242 (100.0%)</li>
<li>Failed: 0/147242 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 147291/147292 (100.0%)</li>
<li><details><summary>Failed: 1/147292 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.binding_model.shader_access.primary_cmd_buf.with_template.combined_image_sampler_mutable.vertex.multiple_descriptor_sets.descriptor_array.cube
Stack trace: Flake
```
</p></details></li>


</ul></details>


</ul></details>

**Comment by amdrexu** [comment]
There are code conflicts. Maybe need a rebase.

**Comment by JaxLinAMD** [comment]
right, @perlfu , please resolve conflicts.

**Pull request merged**